### PR TITLE
Proposal: Split transactions endpoint by group

### PR DESCRIPTION
**Pull request by dgarcia360**
## Problem

Reported by @rg911 
> The new searchTransaction returns the aggregate transaction without inner transactions. Then the transaction/{id} only returns confirmed transactions. It is not possible to get an aggregate bonded cosigned by purely using the rest api.

## Alternatives discussed

**A)  Pass the group as a path param.**

New endpoints:  ``/transactions/:group/`` and ``/transactions/:group/:id|hash``

(+) Mongo stores each group as a separate collection.
(+)  Aggregate transactions can be cosigned by hash, without having to request all the aggregate transactions related to the account.
(+)  Custom filters can be added to a specific group in future iterations (e.g. return only partial transactions not signed or return only confirmed transactions finalized).
(-) Multiple endpoints.

**B) Catapult-server merges transactions under the same collection.**

(+) One endpoint approach.
(-)  Server has good reasons to not merge all the collections: Pruning partial and unconfirmed txs at different intervals.

**C) REST queries all the transaction-related collections at the same time.**

(+) One endpoint approach.
(-) Ids between collections could cause conflicts.

**Solution proposed : A**

## Other changes

Decided with the REST team to rename ``transactions/:hash/status`` -> ``/transactionStatus`` because:
* It does not return transactions but status.
* Enters in conflict with the new set of endpoints.
* REST defines ``transactionStatus`` as a separate set of routes.

Alternatives evaluated: ``transactions/status/:hash``

### CODE REVIEW

File: spec/core/transaction/routes/getTransactionConfirmed.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+tags:
+  - Transaction routes
+summary: Get transaction information
+description: Returns confirmed transaction information given a transactionId or hash.
+operationId: getTransactionConfirmed
```
[the remark this answers is not in the document]
Comment (dgarcia360) [reply spec/core/transaction/routes/getTransactionConfirmed.yml:5]: Agree - fixed!

**Comment by Vektrat** [comment]
> Would it better to have `group / status` as a path param in `transactions` endpoint rather than using 3 separate. Server side can still call different collections.

we will treat as a param server side so it doesn't really matter to us how openapi represents it

**Comment by dgarcia360** [comment]
I decided to separate them in order to allow:

(+) Custom filters can be added to a specific group in future iterations (e.g. return only partial transactions not signed or return only confirmed transactions finalized).

This would permit to solve this issue https://github.com/nemtech/catapult-rest/issues/244

If you think we are not going to need it, I'm ok with adding a path param based on the enum.

**Comment by rg911** [comment]
> I decided to separate them in order to allow:
> 
> (+) Custom filters can be added to a specific group in future iterations (e.g. return only partial transactions not signed or return only confirmed transactions finalized).
> 
> This would permit to solve this issue [nemtech/catapult-rest#244](https://github.com/nemtech/catapult-rest/issues/244)
> 
> If you think we are not going to need it, I'm ok with adding a path param based on the enum.

Either way will do for me

**Comment by Vektrat** [comment]
> Why not using a path parameter instead of having each endpoint 3 times?

it's explained above

**Comment by fboucquez** [comment]
One issue related to splitting the endpoints by state is that the user would need to know the status of the transaction, not only the id in order to retrieve the data. Also, the status the user knew could not be valid anymore, for example, the transaction has been moved from unconfirmed to confirmed, the unconfirmed endpoint by id would return 404.

**Comment by Vektrat** [comment]
> One issue related to splitting the endpoints by state is that the user would need to know the status of the transaction, not only the id in order to retrieve the data. Also, the status the user knew could not be valid anymore, for example, the transaction has been moved from unconfirmed to confirmed, the unconfirmed endpoint by id would return 404.

- the proposal is exactly the same paradigm that was offered by the API before we suggested a badly designed one, so i don't think it is so dramatic to go back on this we can always decide if we want to improve this afterwards, and specially now, where we are in a hurry to move forward with this change
- we won't be able to paginate over all the transactions as one full group as things are now for the motives already mentioned multiple times in multiple discussions in the very recent past

**Comment by fboucquez** [comment]
It's ok, I stated my points. If you don't agree. It's fine.

**Comment by Vektrat** [comment]
> fine

i'm not saying i don't agree, they are facts, and they have been considered when deciding that, but we can't have everything at no loss

and you didn't just state your points, you also said `pagination is broken`, when it isn't, you just don't like the way in which it was designed, and also, this issue has _nothing_ to do with pagination, it's about merging different collections/endpoints

**Comment by fboucquez** [comment]
I do like the new design, especially the user query params in order to mix and match options.  I don't like splitting objects by state at API level. 

If you have 100 confirmed, 100 unconfirmed, and 100 partials and you want to show them in pages of 50 from newest to oldest, how would you do it?

**Comment by Vektrat** [comment]
i'm just saying that this is staying just as it was before, and that pagination _is_ working, this is related to the redesign of endpoints, which in the end wasn't as easy as it was thought, so we aren't going back on anything, we are still moving forward, and we can still consider this in the future once 9.6 deadline is met

there are many things that can't be done with this and many other APIs, we _know_ the current issues but there are many more things on the table that *you* have probably not considered but you just think it is as easy as doing a join - first of all, the real meaning of the statuses that david has been trying to explain to you, second, the architectural constraints that i won't get my hands dirty with, basically because the database design and administration is NOT under REST's hands as it usually is in most projects,  third, assuming we could join collections, we would lose temporality, because in most databases when doing so you can still sort by other factors or timestamps, but in our situation, we can't, so we would be relying on yet another complex sorting which would probably not be fully consistent either, and we can go on - but most importantly, we committed to having 9.6 ready next week, and i doubt anyone wants such changes on REST/openapi, one day before delivery

**Comment by fboucquez** [comment]
That's why I said it is fine. Let's move on, we will be creating the new endpoints in the SDKs. Using path param would have simplified this side of the API but It's ok if the params may change in the future.

**Comment by Vektrat** [comment]
we expect/hope to have those changes by tomorrow

**Pull request merged**